File: camaleonte/camaleonte/metasploit/camaleonte/util/protocols/protocol.rb
require 'time'

require_relative '../utils'
require_relative '../mediums/filesystem'

CONNECTION_POLL_DELAY = 0.1

module Protocol
  attr_reader :filesystem

  def initialize(filesystem)
    @filesystem = filesystem
  end

  # Abstract methods - should be overridden in subclass
  def encode_file(file, data)
    raise NotImplementedError, "#{self.class} must implement encode_file"
  end

  def decode_file(file)
    raise NotImplementedError, "#{self.class} must implement decode_file"
  end

  def data_per_file
    raise NotImplementedError, "#{self.class} must implement data_per_file"
  end

  # Initial connection methods
  def connect
    current_count = @filesystem.get_client_count
    @filesystem.set_client_count(current_count + 1)
    @filesystem.set_channel_pos(current_count)
    @filesystem.update_virtual_filesystem
    @filesystem.set_signal(Signal::CLEAR)
  end

  def wait_for_connection
    current_count = @filesystem.get_client_count
    sleep CONNECTION_POLL_DELAY while current_count == @filesystem.get_client_count
    @filesystem.set_channel_pos(current_count)
    @filesystem.update_virtual_filesystem
    @filesystem.set_signal(Signal::CLEAR)
  end

  # Transmission methods
  def read
    data = ''.b
    loop do
      # wait for a DONE signal
      while @filesystem.read_signal != Signal::DONE
      end
      # read curent batch
      current_batch = ''.b
      @filesystem.get_files.each do |file|
        current_batch << decode_file(file)
        break if current_batch.include?(TERMINATOR)
      end
      # puts "RECEIVED BATCH: #{current_batch.length}" # DEBUG
      # puts current_batch.inspect # DEBUG
      received_hash = current_batch.byteslice(0, CHECKSUM_HASH_SIZE)
      calculated_hash = checksum_hash(current_batch.byteslice(CHECKSUM_HASH_SIZE..-1))
      if received_hash == calculated_hash
        data << current_batch.byteslice(CHECKSUM_HASH_SIZE..-1)
        @filesystem.set_signal(Signal::ACK)
      else
        @filesystem.set_signal(Signal::NACK)
      end

      if data.include?(TERMINATOR)
        data = data.split(TERMINATOR, 2)[0]
        break
      end
    end
    data
  end

  def write(data)
    # wait until signal cleared
    while @filesystem.read_signal != Signal::CLEAR
    end

    payload = data + TERMINATOR
    files = @filesystem.get_files
    total_files = files.size

    data_per_batch = data_per_file * total_files - CHECKSUM_HASH_SIZE
    raise 'NOT ENOUGH FILES' if data_per_batch <= 0

    batches = []
    i = 0
    while i < payload.bytesize
      batches << payload.byteslice(i, data_per_batch)
      i += data_per_batch
    end

    cur_batch_i = 0
    while cur_batch_i < batches.size
      batch = batches[cur_batch_i]
      # puts "Sent Hash: #{checksum_hash(batch).inspect}" # DEBUG
      batch = checksum_hash(batch) + batch

      file_chunks = []
      i = 0
      while i < batch.bytesize
        file_chunks << batch.byteslice(i, data_per_file)
        i += data_per_file
      end

      file_chunks.each_with_index do |chunk, idx|
        # puts "writing #{chunk.inspect} to #{files[idx]}" # DEBUG
        encode_file(files[idx], chunk)
      end

      # puts "SENT BATCH: #{batch.bytesize}" # DEBUG
      # puts batch.inspect # DEBUG
      @filesystem.set_signal(Signal::DONE)

      loop do
        sig = @filesystem.read_signal
        if sig == Signal::ACK
          cur_batch_i += 1
          break
        elsif sig == Signal::NACK
          break
        end
      end
    end
    @filesystem.set_signal(Signal::CLEAR)
  end
end

File: camaleonte/camaleonte/python-cc/src/protocol/protocol.py
import time
from abc import ABC, abstractmethod

from src.mediums.filesystem import Filesystem, Signal
from src.utils import TERMINATOR, CHECKSUM_HASH_SIZE, checksum_hash


CONNECTION_POLL_DELAY = .1

# TODO: add method to pause and recalculate batches when new client joins (VFS change)
class Protocol(ABC):
    def __init__(self, filesystem: Filesystem) -> None:
        self.filesystem = filesystem


    ### INITIAL CONNECTION
    def connect(self):
        # increment client count
        current_count = self.filesystem.get_client_count()
        self.filesystem.set_client_count(current_count+1)
        # set channel pos to index
        self.filesystem.set_channel_pos(current_count)
        # update vfs + clear sync
        self.filesystem.update_virtual_filesystem()
        self.filesystem.set_signal(Signal.CLEAR)


    def wait_for_connection(self):
        # wait for count to be incremented
        current_count = self.filesystem.get_client_count()
        while current_count == self.filesystem.get_client_count():
            time.sleep(CONNECTION_POLL_DELAY)
        # set the channel pos
        self.filesystem.set_channel_pos(current_count)
        # update vfs + clear sync
        self.filesystem.update_virtual_filesystem()
        self.filesystem.set_signal(Signal.CLEAR)

    ### READ/WRITE
    def read(self) -> bytes:
        data = b''
        # keep on reading until terminator found
        while True:
            # wait for a done signal
            while self.filesystem.read_signal() != Signal.DONE:
                pass
            print("[READ] DONE")
            # read the current batch
            current_batch = b''
            for file in self.filesystem.get_files():
                current_batch += self.decode_file(file)
                if TERMINATOR in current_batch:
                    break
            print("RECEIVED BATCH:", len(current_batch))
            print(current_batch)
            # verify the batch
            received_hash = current_batch[0:CHECKSUM_HASH_SIZE]
            calculated_hash = checksum_hash(current_batch[CHECKSUM_HASH_SIZE:])
            # if the hash is correct
            if received_hash == calculated_hash:
                data += current_batch[CHECKSUM_HASH_SIZE:]
                self.filesystem.set_signal(Signal.ACK)
            else:
                self.filesystem.set_signal(Signal.NACK)
            # check if we are done reading
            if TERMINATOR in data:
                data = data.split(TERMINATOR)[0]
                break
        return data

    def write(self, data: bytes) -> None:
        # ensure that signal is cleared
        while self.filesystem.read_signal() != Signal.CLEAR:
            pass
        # split up into "batches" or "packets"
        payload = data + TERMINATOR
        files = self.filesystem.get_files()
        total_files = len(files)
        # find if valid file count and amnt of data per batch
        DATA_PER_BATCH = self.data_per_file() * total_files - CHECKSUM_HASH_SIZE
        if DATA_PER_BATCH <= 0:
            Exception("NOT ENOUGH FILES")
        # split up data into batches
        batches = []
        for i in range(0, len(payload), DATA_PER_BATCH):
            batches.append(payload[i:i + DATA_PER_BATCH])
        # send all the batches
        cur_batch_i = 0
        while cur_batch_i < len(batches):
            # add the checksum to beginning
            batch = batches[cur_batch_i]
            print("Sent Hash:", checksum_hash(batch))
            batch = checksum_hash(batch) + batch
            # split up the batch into file sized chunks
            file_chunks = []
            for i in range(0, len(batch), self.data_per_file()):
                file_chunks.append(batch[i:i+self.data_per_file()])
            # write each file chunk
            for i, chunk in enumerate(file_chunks):
                self.encode_file(files[i], chunk)
            # tell receiver that we are done writing batch
            print("SENT BATCH:", len(batch))
            print(batch)
            self.filesystem.set_signal(Signal.DONE) 
            # wait for ACK or NACK
            while True:
                sig = self.filesystem.read_signal()
                if sig == Signal.ACK:
                    print("[READ] ACK")
                    cur_batch_i += 1
                    break
                if sig == Signal.NACK:
                    print("[READ] NACK")
                    break
        # done writing all batches
        self.filesystem.set_signal(Signal.CLEAR)

    ### THESE METHODS NEED TO BE IMPLEMENTED
    @abstractmethod
    def encode_file(self, file: str, data: bytes) -> None:
        pass

    @abstractmethod
    def decode_file(self, file: str) -> bytes:
        pass

    @abstractmethod
    # returns amount of bytes can be encoded per file
    def data_per_file(self) -> int:
        pass

    